Validate the default JAXBContext at build time only if it is really u…

ppalaga opened the pull request:
…sed in the application; do not validate if user provides his own JAXBContext bean or if there is no JAXBContext injection point

Fix #31646

ppalaga committed:
Validate the default JAXBContext at build time only if it is really used
in the application; do not validate if user provides his own JAXBContext
bean or if there is no JAXBContext injection point #31646

Sgitario commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:
```diff
@@ -297,29 +297,55 @@ void registerClasses(
+        final BeanResolver beanResolver = beanContainerState.getBeanResolver();
+        final Set<BeanInfo> beans = beanResolver
+                .resolveBeans(Type.create(DotName.createSimple(JAXBContext.class), org.jboss.jandex.Type.Kind.CLASS));
+        if (config.validateJaxbContext && !beans.isEmpty()) {
```
I don't know what is the performance penalty of calling `beanResolver.resolveBeans`, but we should only discover the beans if `validateJaxbContext` is true.

ppalaga replied:
+1, let me reorder the conditions

Sgitario commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:
```diff
@@ -297,29 +297,55 @@ void registerClasses(
+
+        final BeanResolver beanResolver = beanContainerState.getBeanResolver();
+        final Set<BeanInfo> beans = beanResolver
+                .resolveBeans(Type.create(DotName.createSimple(JAXBContext.class), org.jboss.jandex.Type.Kind.CLASS));
```
Missing the Marshaller and Unmarshaller types too.

ppalaga replied:
Marshaller and Unmarshaller require the JAXBContext bean to exist, so checking their existence is not necessary IMO.  Let me add a test to prove that.

ppalaga replied:
There is now ConflictingModelClassesMarshalerOnlyTest to cover the situation where there conflicting classes, no JAXBContext injection point, but there is a Marshaller injection point. The application fails on boot as expected.

Sgitario commented on extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:
```diff
@@ -297,29 +297,55 @@ void registerClasses(
+                try {
+                    JAXBContext.newInstance(filteredClassesToBeBound.getClasses().toArray(new Class[0]));
+                } catch (JAXBException e) {
+                    throw new IllegalStateException("Failed to configure JAXB context", e);
```
Produce a ValidationErrorBuildItem build item instead of throwing this exception.

ppalaga replied:
Good idea, let me do that. 
I just copied the old code without too much thinking about it. I had to add the ValidationErrorBuildItem consumer to the method signature to force the build time container to invoke the method.

ppalaga replied:
I had to remove the ValidationErrorBuildItem producer from the method signature because of the dependency cycle in resteasy-reactive-jaxb. Now the DeploymentException is thrown directly from validateDefaultJaxbContext() much like it was before this change.

Sgitario replied:
ok ok!

Sgitario reviewed (changes requested):
Apart from the comments, these changes cause a build cycle exception in the RESTEasy Reactive Jaxb tests.

ppalaga commented:
5e39264: 

* Addressed the feedback by @Sgitario (thanks!)
* Solved the BuildStep methods cycle by not producing a ValidationErrorBuildItem and rather throwing a DeploymentException direcly plus consuming a JaxbContextConfigRecorder to force the container to call validateDefaultJaxbContext().

@Sgitario please feel free to re-review

Sgitario reviewed (approved):
lgtm! thanks @ppalaga !

btw, I still think https://github.com/quarkusio/quarkus/pull/31651 is something we could incorporate. Let me know if you want to add this change as part of your pull request or I should add it after yours is merged.

ppalaga commented:
> I still think https://github.com/quarkusio/quarkus/pull/31651 is something we could incorporate.

On one hand, it looks like a nice general improvement. 
OTOH, I wonder how transparent can we make it for the users to understand that in spite they have a dependency bringing org.acme.Foo class, that class might still not be available in the default JAXBContext because some extension "Bar" that the user also happens to have in his classpath chose to exclude org.acme.Foo? However well we document that behavior in extension "Bar", it can still be very surprising to the users. 

Having a catch-all default JAXBContext is definitely intuitive and user friendly, but it is very hard to say who should be responsible for solving the namespace conflicts. I do not feel like it should (always) be extension devs, because they might not know what the users want. Maybe sometimes the choice is clear, e.g. when one half of the conflicting classes is deprecated or outdated. Maybe in such situations `ExcludeJaxbClassesToBeBoundBuildItem` would be useful.
